File: WoofIOS/SwiftPackages/Domain/UseCase/Sources/UseCase/Woofs/HasWoofedDogUseCase.swift
import Combine
import DI
import Models
import Repositories

@Factory
public final class HasWoofedDogUseCase {
    private let woofsRepository: WoofsRepository

    public func callAsFunction(dog: Dog) -> AnyPublisher<Bool, Never> {
        woofsRepository.woofedDogIds
            .map { $0.contains(dog.id) }
            .removeDuplicates()
            .eraseToAnyPublisher()
    }
}

File: WoofAndroid/domain/usecase/src/main/kotlin/com/perrystreet/woof/usecase/woofs/HasWoofedDogUseCase.kt
package com.perrystreet.woof.usecase.woofs

import com.perrystreet.woof.models.dog.Dog
import com.perrystreet.woof.repositories.woofs.WoofsRepository
import io.reactivex.rxjava3.core.Observable
import org.koin.core.annotation.Factory

@Factory
class HasWoofedDogUseCase(
    private val woofsRepository: WoofsRepository,
) {
    operator fun invoke(dog: Dog): Observable<Boolean> =
        woofsRepository.woofedDogIds
            .map { dog.id in it }
            .distinctUntilChanged()
}
